### ISSUE TITLE
NaN in training

### ISSUE THREAD

**Issue by yd-yin** [kerrj/lerf#3]
Hi, thanks for your excellent work! The project is really cool!

However, I've encountered NaN in training, and the scene in viewer becomes completely black. 

I followed the instructions to use `lerf-merge` branch of nerfstudio, and use the demo `poster` dataset. I got reasonable results with nerfacto, but got NaN with lerf. 

Do you have any idea what the problem is? Thanks a lot.

![image](https://user-images.githubusercontent.com/37136395/231192918-e3f613b0-4df1-4e25-9a41-2e3ae1bb0cae.png)
![image](https://user-images.githubusercontent.com/37136395/231193183-b7fc40ae-5929-4c0f-b0f7-2c1d31ffac24.png)

**Comment by kerrj** [comment]
Thanks for opening an issue!
This is also an issue we've noticed since refactoring code: on some scenes the training can randomly NaN out anywhere between 2k and 10k steps (or not at all!), it seems quite random. There are a few potential fixes which we haven't had time to try yet (working on viewer refactor) but it'd be great if you had some time to investigate :)

1. Currently the loss uses huber_loss out of the box, which could be replaced with a torch.nanmean(huber_loss(... reduction='none')). This might solve it by ignoring nans in the loss entirely
2. If this doesn't work, playing with optimizer parameters could get some mileage; for example lowering the language head learning rate, or more sharply decaying the learning rate.

**Comment by yd-yin** [comment]
Hi, thanks for your quick response!

Well, it would be fun to investigate this issue, but before that, I think it would also be good if I can see a final result. (So far I did not get a single worked scene:))

Would you like to offer some more details on how to reproduce a good result, say, the teaser? E.g., the data, hyperparameters, even if random seed if necessary. If the adaptation to some specific UI or viewer can be a barrier, I think it's totally OK to use a vanilla and less fancy one.

**Comment by kerrj** [comment]
Have you tried using the viewer to input text? you should be able to see reasonable results within ~1-2k steps if you play around with text queries in the viewer. You can also download datasets from our website from the "data" link or [here](https://drive.google.com/drive/folders/1vh0mSl7v29yaGsxleadcj-LCZOE_WEWB?usp=share_link).

**Comment by kerrj** [comment]
[This commit](https://github.com/kerrj/lerf/commit/92ed947d881a11975226165cadfd44f949e0e753) could help, it's a bit of a band-aid but it does seem to prevent training from collapsing. I think it might have some other side effects though, so I'm still investigating optimizer parameters as a better alternative.

**Comment by kerrj** [comment]
Adding weight decay also seems to have helped a lot, the latest code should hopefully improve this issue quite a bit.

**Comment by kerrj** [comment]
Ok great! closing this for now. I think the reason for NaNs is probably because we're using large vectors with the hashgrid along with mixed precision training, which can be a bit unstable.

**Comment by yd-yin** [comment]
Thanks!

**Comment by krrish94** [comment]
Here's another potential reason for NaN outputs that I encountered in another application, but useful to explore here.

When interpolating CLIP embeddings across scales, sometimes, we can end up with nans in some dimensions of (or an entire) interpolated CLIP embedding. Another implementation detail of note: the codebase seems to use calls to `.norm(p=2, dim=-1, keepdim=True)` to normalize embeddings, which can cause issues when embedding norms are quite close to zero. (A solution here is to use `torch.nn.functional.normalize(<NAME_OF_TENSOR>, p=2, dim=-1, eps=1e-12)`, which implicitly deals with division by zero-norm vectors).

One quick and dirty fix is to find all nan dimensions and set them to zero, right after we've interpolated in line https://github.com/kerrj/lerf/blob/c269778f2148452ed5d221130a21c9f3ce91cb85/lerf/data/lerf_datamanager.py#L115
(A more rigorous workaround could be to flag these embeddings altogether, and remove them from consideration when computing the loss).

I tried this (with good success) in a different application. That isn't to say it should work here too, but FWIW, this could help others encountering NaNs during training
